pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/AbstractPartitionAwareRoutingTableBuilder.java

```diff
@@ -0,0 +1,120 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.broker.routing.builder;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.commons.configuration.Configuration;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linkedin.pinot.broker.pruner.SegmentPrunerContext;
+import com.linkedin.pinot.broker.pruner.SegmentZKMetadataPrunerService;
+import com.linkedin.pinot.broker.routing.RoutingTableLookupRequest;
+import com.linkedin.pinot.common.config.TableConfig;
+import com.linkedin.pinot.common.metadata.segment.SegmentZKMetadata;
+import com.linkedin.pinot.common.response.ServerInstance;
+import com.linkedin.pinot.transport.common.SegmentId;
+import com.linkedin.pinot.transport.common.SegmentIdSet;
+
+
+/**
+ * Abstract partition aware routing table builder for the Kafka low level consumer.
+ *
+ * For an external view change, a subclass is in change of updating the look up table that is used
+ * for routing. The look up table is in the format of < segment_name -> (replica_id -> server_instance) >.
+ *
+ * When the query comes in, the routing algorithm is as follows:
+ *   1. Randomly pick a replica id (or replica group id)
+ *   2. For each segment of the given table,
+ *      a. Check if the segment can be pruned. If pruned, go to the next segment.
+ *      b. If not pruned, assign the segment to a server with the replica id that is picked above.
+ *
+ */
+public abstract class AbstractPartitionAwareRoutingTableBuilder extends AbstractRoutingTableBuilder {
+  private static final Logger LOGGER = LoggerFactory.getLogger(AbstractPartitionAwareRoutingTableBuilder.class);
+
+  protected static final String PARTITION_METADATA_PRUNER = "PartitionZKMetadataPruner";
+  protected static final int NO_PARTITION_NUMBER = -1;
+
+  Map<SegmentId, Map<Integer, ServerInstance>> _segmentNameToServersMapping = new HashMap<>();
+  AtomicReference<Map<SegmentId, Map<Integer, ServerInstance>>> _mappingReference =
+      new AtomicReference<>(_segmentNameToServersMapping);
+  Map<SegmentId, SegmentZKMetadata> _segmentToZkMetadataMapping = new ConcurrentHashMap<>();
+  protected ZkHelixPropertyStore<ZNRecord> _propertyStore;
+  protected SegmentZKMetadataPrunerService _pruner;
+  protected TableConfig _tableConfig;
+  protected Random _random = new Random();
+  protected int _numReplicas;
+
+  @Override
+  public void init(Configuration configuration, TableConfig tableConfig, ZkHelixPropertyStore<ZNRecord> propertyStore) {
+    _tableConfig = tableConfig;
+    _propertyStore = propertyStore;
+    _pruner = new SegmentZKMetadataPrunerService(new String[]{PARTITION_METADATA_PRUNER});
```

Comment:
Not required for this PR, but what pruners to wire-in should come in via config, as opposed to being hard coded here.